otel-android/src/main/java/com/solarwinds/android/SessionIdBasedSampler.java

```diff
@@ -0,0 +1,74 @@
+package com.solarwinds.android;
+
+import androidx.annotation.NonNull;
+import io.opentelemetry.android.session.SessionProvider;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.samplers.Sampler;
+import io.opentelemetry.sdk.trace.samplers.SamplingResult;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * A custom OpenTelemetry sampler that makes sampling decisions based on a session ID. The session
+ * ID is provided externally via {@link SessionProvider}.
+ */
+public class SessionIdBasedSampler implements Sampler {
+
+    private final double rate;
+
+    private final double threshold;
+
+    private final SessionProvider sessionProvider;
+
+    public SessionIdBasedSampler(double rate, SessionProvider sessionProvider) {
+        this.rate = rate;
+        this.sessionProvider = sessionProvider;
+        this.threshold = rate * Long.MAX_VALUE;
+    }
+
+    @Override
+    public SamplingResult shouldSample(
+            @NonNull Context parentContext,
+            @NonNull String traceId,
+            @NonNull String name,
+            @NonNull SpanKind spanKind,
+            @NonNull Attributes attributes,
+            @NonNull List<LinkData> parentLinks) {
+
+        String sessionId = sessionProvider.getSessionId();
+        if (!sessionId.isEmpty()) {
+            long hashedValue = hash(sessionId);
+            if (hashedValue < threshold) {
+                return SamplingResult.recordAndSample();
+            }
+        }
+
+        return SamplingResult.drop();
+    }
+
+    @Override
+    public String getDescription() {
+        return String.format(
+                Locale.getDefault(),
+                "SessionIdBasedSampler{rate=%f, threshold=%f}",
+                rate,
+                threshold);
+    }
+
+    private long hash(String sessionId) {
+        byte[] bytes = sessionId.getBytes(StandardCharsets.UTF_8);
+        return hashBytes(bytes);
+    }
+
+    private long hashBytes(byte[] bytes) {
+        long hash = 98764321261L;
+        for (byte bite : bytes) {
+            hash = (hash * 343443379) ^ (bite & 0xff);
+        }
+        return hash & Long.MAX_VALUE;
```

Comment:
Two questions
1. Given this is a FNV hash, why not use known FNV primes to improve dispersion ? Very minor but I don't see any reason not to.
2. Why `bite & 0xff` ? Wouldn't that always be the same as just `bite` or is there some Java semantics I'm not aware of at play ?

Reply:
> Given this is a FNV hash, why not use known FNV primes to improve dispersion ? Very minor but I don't see any reason not to.

It's an abridged murmur hash. Do you have any FNV prime numbers in my mind, I could use?

> Why bite & 0xff ? Wouldn't that always be the same as just bite or is there some Java semantics I'm not aware of at play ?
nope, that was unnecessary.

Reply:
it looks like `14695981039346656037`  is too big, have something smaller that fits in a `long`?